First iteration

altenratten opened the pull request.

altenratten committed:
Реализован Шаг 3. Правила повторения задач. Ошибок на тестах нет

altenratten committed:
Добавлен task.go. В нём type Task struct и func AddTask

altenratten committed:
=== RUN   TestAddTask
--- PASS: TestAddTask (0.13s)
PASS

altenratten committed:
Шаг 5. Получаем список ближайших задач Passed
=== RUN   TestTasks
--- PASS: TestTasks (0.16s)
PASS
ok      github.com/altenratten/go_final_project/tests   0.620s

altenratten committed:
Шаг 6. Редактирование задачи Passed
=== RUN   TestEditTask
ok      github.com/altenratten/go_final_project/tests   0.556s

altenratten committed:
Шаг 7. Заканчиваем реализацию API
=== RUN   TestAddTask
--- PASS: TestAddTask (0.14s)
=== RUN   TestApp
--- PASS: TestApp (0.13s)
=== RUN   TestDB
--- PASS: TestDB (0.06s)
=== RUN   TestNextDate
--- PASS: TestNextDate (0.01s)
=== RUN   TestTask
--- PASS: TestTask (0.02s)
=== RUN   TestEditTask
--- PASS: TestEditTask (0.04s)
=== RUN   TestDone
--- PASS: TestDone (0.10s)
=== RUN   TestDelTask
--- PASS: TestDelTask (0.03s)
=== RUN   TestTasks
--- PASS: TestTasks (0.11s)
PASS
ok      github.com/altenratten/go_final_project/tests   1.525s

altenratten committed:
Project complite

altenratten committed:
delete Token

altenratten committed:
Добавил реализацию Docker

altenratten committed:
Change Port to 8080

altenratten committed:
Change README

altenratten committed:
modified:   README.md

altenratten committed:
modified:   README.md

altenratten committed:
modified:   README.md

dirtymew commented on Dockerfile:
```diff
@@ -0,0 +1,41 @@
+COPY --from=builder /app/server /usr/local/bin/server
+
+# Expose port
+EXPOSE 80
```
а что будет если TODO_PORT поменяется

dirtymew commented on README.md:
```diff
@@ -1,5 +1,71 @@
 # Файлы для итогового задания
 
+Веб-сервер, который реализует функциональность планировщика задач — это аналог TODO-листа
```
супер что добавил документацию

dirtymew commented on pkg/api/api.go:
```diff
@@ -0,0 +1,35 @@
+
+// Init — initializes the API
+func Init() {
+	http.HandleFunc("/api/task", auth(taskHandler))
```
добавил авторизацию - отлично

но нужно включить опцию в тестах добавив Token

dirtymew commented on pkg/api/nextdate.go:
```diff
@@ -0,0 +1,106 @@
+	nowStr := r.FormValue("now")
+	dateStr := r.FormValue("date")
+	repeat := r.FormValue("repeat")
+
```
методы не во всех хендлерах проверяются - нужно поправить

dirtymew commented on pkg/api/writeJson.go:
```diff
@@ -0,0 +1,12 @@
+// writeJSON — helper function to return JSON response
+func writeJson(w http.ResponseWriter, data any) {
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+	json.NewEncoder(w).Encode(data)
```
ошибка из Encode не обработана

dirtymew commented on pkg/api/task.go:
```diff
@@ -0,0 +1,113 @@
+
+	task, err := db.GetTask(id)
+	if err != nil {
+		writeJson(w, map[string]string{"error": "task not found"})
```
writeJson не учитывает код ответа

при ошибка он может быть разный

dirtymew replied:
нужно везде исправлять

dirtymew commented on pkg/db/task.go:
```diff
@@ -0,0 +1,134 @@
+
+// Tasks — gets tasks from the DB
+func Tasks(limit int, search string) ([]*Task, error) {
+	const baseLimit = 50
```
в глобальную константу

dirtymew commented on pkg/db/task.go:
```diff
@@ -0,0 +1,134 @@
+			return nil, err
+		}
+		tasks = append(tasks, &t)
+	}
```
нужно еще ошибку из rows обработать
http://go-database-sql.org/errors.html

dirtymew commented on pkg/db/task.go:
```diff
@@ -0,0 +1,134 @@
+	if err != nil {
+		return "", err
+	}
+	return t.Format("20060102"), nil
```
20060102 в глобальную константу и везде использовать

dirtymew commented on pkg/db/task.go:
```diff
@@ -0,0 +1,134 @@
+// GetTask — gets a task from the DB
+func GetTask(id string) (*Task, error) {
+	row := db.QueryRow("SELECT id, date, title, comment, repeat FROM scheduler WHERE id = ?", id)
+	var t Task
```
а почему сразу не поинтер?

dirtymew commented on pkg/server/server.go:
```diff
@@ -0,0 +1,30 @@
+	fs := http.FileServer(http.Dir("./web"))
+	http.Handle("/", fs)
+
+	log.Printf("Server started on port %s...", port)
```
отличный лог

dirtymew commented on main.go:
```diff
@@ -0,0 +1,36 @@
+	}
+
+	// Initialize the DB
+	if err := db.Init(dbFile); err != nil {
```
по завершению работы нужно закрыть соединение

dirtymew commented on pkg/api/signin.go:
```diff
@@ -0,0 +1,104 @@
+// Middleware for authentication
+func auth(next http.HandlerFunc) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		pass := os.Getenv("TODO_PASSWORD")
```
так делать не стоит в хендлерах

лучше один раз считывать конфигурацию на старте и потом ее использовать

dirtymew reviewed (changes requested):
Привет  Александр

хорошая работа, не плохая структура решения, задания со * и красивая история коммитов)


дал комментарии что поправить и улучшить

после исправления готов принять работу

altenratten committed:
Добавил:
1. Переменная в Dockerfile EXPOSE ${TODO_PORT}
2. Token в settings.go
3. Добавил проверку методов в хендлерах
4. Добавил обработку ошибок из Encode в func writeJson
5. Добавил в writeJson код ответа. Добавил везде коды.
6. Вынес const  в глобальные
7. Обработал ошибку из rows
8. Сокртил GetTask в db/task.go
9. Добавил закрытие соединение с БД по завершению
10. Вынес var pass из хендлера

altenratten committed:
modified:   README.md

altenratten commented:
Добавил коммит с исправлениями

dirtymew reviewed (approved):
Александр

работа принята

поздравляю и желаю удачи
